pennylane/transforms/op_transforms.py

```diff
@@ -0,0 +1,481 @@
+            return args[0]
+
+        return super().__new__(cls)
+
+    def __init__(self, fn):
+        if not callable(fn):
+            raise OperationTransformError(
+                f"The operator function to register, {fn}, "
+                "does not appear to be a valid Python function or callable."
+            )
+
+        self._fn = fn
+        self._sig = inspect.signature(fn).parameters
+        self._tape_fn = None
+        functools.update_wrapper(self, fn)
+
+    def __call__(self, *targs, **tkwargs):
+        obj = None
+
+        if targs:
+            obj, *targs = targs
+
+        if isinstance(obj, (qml.operation.Operator, qml.tape.QuantumTape)) or callable(obj):
+            return self._create_wrapper(obj, *targs, **tkwargs)
+
+        # Input is not an operator nor a QNode nor a quantum tape nor a qfunc.
+        # Assume Python decorator syntax:
+        #
+        # result = op_func(*transform_args)(obj)(*obj_args)
+        #
+        # or
+        #
+        # @op_func(*transform_args)
+        # @qml.qnode(dev)
+        # def circuit(...):
+        #     ...
+        # result = circuit(*qnode_args)
+
+        # Prepend the input to the transform args,
+        # and create a wrapper function.
+        if obj is not None:
+            targs = (obj,) + tuple(targs)
+
+        def wrapper(obj):
+            return self._create_wrapper(obj, *targs, **tkwargs)
+
+        return wrapper
+
+    def fn(self, obj, *args, **kwargs):
+        """Evaluate the underlying operator transform function.
+
+        If a corresponding tape transform for the operator has been registered
+        using the :attr:`.op_transform.tape_transform` decorator,
+        then if an exception is raised while calling the transform function,
+        this method will attempt to decompose the provided object for the tape
+        transform.
+
+        Args:
+            obj (.Operator, pennylane.QNode, .QuantumTape, or Callable): An operator, quantum node, tape,
+                or function that applies quantum operations.
+            *args: positional arguments to pass to the function
+            **kwargs: keyword arguments to pass to the function
+
+        Returns:
+            any: the result of evaluating the transform
+        """
+        try:
+            return self._fn(obj, *args, **kwargs)
+
+        except Exception as e1:  # pylint: disable=broad-except
+
+            try:
+                # attempt to decompose the operation and call
+                # the tape transform function if defined
+                return self.tape_fn(obj.expand(), *args, **kwargs)
+
+            except (AttributeError, OperationTransformError):
+                # if obj.expand() does not exist, or the tape transform
+                # function does not exist, simply raise the original exception
+                raise e1 from None
+
+    def tape_fn(self, obj, *args, **kwargs):
+        """Evaluate the underlying tape transform function.

```

Comment:
How about having `evaluate` in the name here? This seems to differ from the `register` functionality of let's say the `tape_transform` method, yet the stark difference is not reflected in the name of the methods.

Reply:
I think the docstring is incorrect 🤔 there is no evaluation happening here, this is just a wrapper around `tape_fn`.
```suggestion
    def tape_fn(self, obj, *args, **kwargs):
        """The underlying tape transform function.
```